aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -840,6 +842,91 @@ Tensor index_select_cuda(const Tensor& self, int64_t dim, const Tensor& index) {
   return out;
 }
 
+template<typename T>
+struct NonZeroOp
+{
+    __host__ __device__ __forceinline__ bool operator()(const T& a) const {
+      return (a!=T(0));
+    }
+};
+
+template<typename scalar_t>
+void nonzero_cuda_out_impl(const Tensor& self, Tensor& out){
+  Tensor self_ = self.contiguous();
+  int N = self_.numel();
+  //compute number of nonzero elements
+  size_t temp_storage_bytes=0;
+  auto& allocator = *c10::cuda::CUDACachingAllocator::get();
+  auto num_nonzeros = allocator.allocate(sizeof(int));
+  cub::TransformInputIterator<bool, NonZeroOp<scalar_t>, scalar_t*> itr(self_.data_ptr<scalar_t>(), NonZeroOp<scalar_t>());
+  const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  cub::DeviceReduce::Sum(nullptr, temp_storage_bytes, itr, (int*)num_nonzeros.get(), N, stream);
+  auto temp_storage = allocator.allocate(temp_storage_bytes);
+  cub::DeviceReduce::Sum(temp_storage.get(), temp_storage_bytes, itr, (int*)num_nonzeros.get(), N, stream);
+  int num_nonzeros_h;
+  C10_CUDA_CHECK(cudaMemcpyAsync(&num_nonzeros_h, num_nonzeros.get(), sizeof(int), cudaMemcpyDeviceToHost, stream));
+  //need to synchronize to make sure data is available on the host
+  C10_CUDA_CHECK(cudaStreamSynchronize(stream));
+  //expected output size is num_nonzeros x ndim
+  //we are producing output with size {num_nonzeros, ndim} and strides {num_nonzeros, 1} (that is, transposed ndim x num_nonzeros output)
+  //we are able to directly use passed output with this size and strides, and we can also (per contract)
+  //resize passed output with incorrect sizes anyway we want.
+  //However, out with correct sizes and incorrect strides will have to be copied to from the intermediate we've produced.
```

Comment:
This is a surprising thing for the out= interface no?
Should we mention it in the `out=` version of the doc that "out must be of the right size and`out.t().is_contiguous()` must be true to avoid any extra allocation" ?
Also I guess this won't be consistent with the CPU version?

Reply:
Yeah, the stride of the output won't be consistent with CPU version. Very few people ever bother to correctly allocate `out` for functions such as `masked_select` or `nonzero` because except in synthetic cases it is never known what the size of the output will be, so it's a very niche case.

Reply:
Note the usage of `nonzero` in advanced indexing https://github.com/pytorch/pytorch/blob/10dd25dcd18457a53e69eb319f48749a49a48430/aten/src/ATen/native/IndexingUtils.h#L36-L39. Given that `select` is along the first dimension, this return format makes indices contiguous, whereas previous cuda (and current cpu) return format makes indices discontiguous. Same for `as_tuple`=True - now cuda returns a tuple of contiguous tensors, previously it was a tuple of discontiguous tensors.

Reply:
On the other hand, I don't see any reason to use out if you don't provide something with what you expect to be of the right size already?
But I agree it's very nice so it should be fine.

Reply:
As we've discussed on other issues, people provide `out` when they are doing manual memory management. If `out` storage is big enough, even if `out` size is not known in advance, the output would still use provided storage.